hummingbot/market/bitfinex/bitfinex_api_user_stream_data_source.py

```diff
@@ -0,0 +1,90 @@
+import asyncio
+import json
+import logging
+import time
+from typing import Optional, List
+
+import ujson
+import websockets
+from websockets.exceptions import ConnectionClosed
+
+from hummingbot.core.data_type.user_stream_tracker_data_source import \
+    UserStreamTrackerDataSource
+from hummingbot.logger import HummingbotLogger
+from hummingbot.market.bitfinex.bitfinex_order_book import BitfinexOrderBook
+from hummingbot.market.bitfinex import BITFINEX_WS_URI
+from hummingbot.market.bitfinex.bitfinex_auth import BitfinexAuth
+from hummingbot.market.bitfinex.bitfinex_order_book_message import \
+    BitfinexOrderBookMessage
+
+
+class BitfinexAPIUserStreamDataSource(UserStreamTrackerDataSource):
+    MESSAGE_TIMEOUT = 30.0
+
+    _logger: Optional[HummingbotLogger] = None
+
+    @classmethod
+    def logger(cls) -> HummingbotLogger:
+        if cls._logger is None:
+            cls._logger = logging.getLogger(__name__)
+        return cls._logger
+
+    def __init__(self, bitfinex_auth: BitfinexAuth, trading_pairs: Optional[List[str]] = None):
+        if trading_pairs is None:
+            trading_pairs = []
+        self._bitfinex_auth: BitfinexAuth = bitfinex_auth
+        self._trading_pairs = trading_pairs
+        self._current_listen_key = None
+        self._listen_for_user_stream_task = None
+        super().__init__()
+
+    @property
+    def order_book_class(self):
+        return BitfinexOrderBook
+
+    async def listen_for_user_stream(self, ev_loop: asyncio.BaseEventLoop, output: asyncio.Queue):
+        while True:
+            try:
+                ws: websockets.WebSocketClientProtocol
+                async with websockets.connect(BITFINEX_WS_URI) as ws:
+                    ws: websockets.WebSocketClientProtocol = ws
+                    payload = self._bitfinex_auth.generate_auth_payload()
+                    await ws.send(json.dumps(payload))
+                    await asyncio.wait_for(ws.recv(), timeout=self.MESSAGE_TIMEOUT)  # info
+                    await asyncio.wait_for(ws.recv(), timeout=self.MESSAGE_TIMEOUT)  # auth
```

Comment:
Considering that we will be processing the message after `_get_response()` and that the response from the initial subscription can be determined, I feel that this is not necessary.

Reply:
removed these rows. Now process them in `_transform_message_from_exchange`